git/cmd.py

```diff
@@ -305,6 +306,173 @@ def __setstate__(self, d: Dict[str, Any]) -> None:
+        # independently of the Windows Git.  A noteworthy example are repos
+        # with Git LFS, where Git LFS may be installed in Windows but not
+        # in WSL.
+        if not is_win:
+            return "bash"
+        try:
+            wheregit = run(["where", Git.GIT_PYTHON_GIT_EXECUTABLE], check=True, stdout=PIPE).stdout
+        except CalledProcessError:
+            return "bash.exe"
+        gitpath = Path(wheregit.decode(defenc).splitlines()[0])
+        gitroot = gitpath.parent.parent
+        gitbash = gitroot / "bin" / "bash.exe"
+        return str(gitbash) if gitbash.exists() else "bash.exe"
+
+    @classmethod
+    def refresh_bash(cls, path: Union[None, PathLike] = None) -> bool:
+        """This gets called by the refresh function (see the top level __init__)."""
+        # Discern which path to refresh with.
+        if path is not None:
+            new_bash = os.path.expanduser(path)
+            new_bash = os.path.abspath(new_bash)
+        else:
+            new_bash = os.environ.get(cls._bash_exec_env_var)
+            if not new_bash:
+                new_bash = cls._get_default_bash_path()
+
+        # Keep track of the old and new bash executable path.
+        old_bash = cls.GIT_PYTHON_BASH_EXECUTABLE
+        cls.GIT_PYTHON_BASH_EXECUTABLE = new_bash
+
+        # Test if the new git executable path is valid. A GitCommandNotFound error is
+        # spawned by us. A PermissionError is spawned if the git executable cannot be
+        # executed for whatever reason.
+        has_bash = False
+        try:
+            run([cls.GIT_PYTHON_BASH_EXECUTABLE, "--version"], check=True, stdout=PIPE)
+            has_bash = True
+        except CalledProcessError:
+            pass
+
+        # Warn or raise exception if test failed.
+        if not has_bash:
+            err = dedent(
+                f"""\
+                Bad bash executable.
+                The bash executable must be specified in one of the following ways:
+                    - be included in your $PATH
+                    - be set via ${cls._bash_exec_env_var}
+                    - explicitly set via git.refresh_bash()
+                """
+            )
+
+            # Revert to whatever the old_bash was.
+            cls.GIT_PYTHON_BASH_EXECUTABLE = old_bash
+
+            if old_bash is None:
+                # On the first refresh (when GIT_PYTHON_GIT_EXECUTABLE is None) we only
+                # are quiet, warn, or error depending on the GIT_PYTHON_REFRESH value.
+
+                # Determine what the user wants to happen during the initial refresh we
+                # expect GIT_PYTHON_REFRESH to either be unset or be one of the
+                # following values:
+                #
+                #   0|q|quiet|s|silence|n|none
+                #   1|w|warn|warning
+                #   2|r|raise|e|error
+
+                mode = os.environ.get(cls._refresh_env_var, "raise").lower()
```

Comment:
I think the mode extraction itself should be factored into its own function and shared. 
It seems like the message itself can be a parameter, along with the environment variable itself.

Reply:
I agree, let's wait until after the review comments were addressed if this is still an issue.